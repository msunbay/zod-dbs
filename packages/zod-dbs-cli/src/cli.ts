import { Command, Option, program } from 'commander';
import { generateZodSchemas } from 'zod-dbs';
import { logDebug, toError, ZodDbsProvider } from 'zod-dbs-core';

import type { ZodDbsCliConfig, ZodDbsCliOptions } from './types.js';

import { enableDebugMode, getConfiguration } from './config.js';
import { importProvider } from './provider.js';
import { getArgumentValue, isSilentMode } from './utils/args.js';
import { logAppName, logError, logSetting } from './utils/logger.js';
import { createProgressHandler } from './utils/progress.js';
import { getAppVersion } from './utils/version.js';

export const runCli = async (cliOptions: ZodDbsCliOptions = {}) => {
  enableDebugMode();
  logDebug('Starting zod-dbs CLI with options:', cliOptions);

  const appVersion = cliOptions.appVersion ?? (await getAppVersion());
  const appName = cliOptions.appName || 'zod-dbs';
  const silent = isSilentMode();

  if (!silent) {
    logAppName(`${appName} CLI v${appVersion}`);
  }

  const config = await getConfiguration(cliOptions);

  program.name(appName);
  program.version(appVersion);
  program.description('Generates Zod schemas from database schema.');

  const provider = await loadProvider(cliOptions.overrides?.provider, config);

  if (!cliOptions.overrides?.provider) {
    program.option('-p,--provider <name>', 'Database provider to use');
  }

  program.option(
    '--config-name <name>',
    'Name of configuration file. E.g. "development" will load "zod-dbs-development.ts"'
  );

  // Add provider-specific dynamic options before parsing the full CLI
  addProviderOptions(program, provider);

  const outputOptions = program.optionsGroup('Output options:');

  const parseBooleanOption = (value: string) => {
    if (value.toLowerCase() === 'true') return true;
    if (value.toLowerCase() === 'false') return false;
    throw new Error(`Invalid boolean value: ${value}`);
  };

  outputOptions.option(
    '-o,--output-dir <path>',
    'Output directory for generated schemas'
  );
  outputOptions.addOption(
    new Option(
      '--module-resolution <type>',
      'Module resolution type for generated files. (defaults to commonjs)'
    ).choices(['commonjs', 'esm'])
  );
  outputOptions.option(
    '--clean-output',
    'Clean output directory before generating schemas'
  );
  outputOptions.option(
    '--exclude <regex>',
    'Exclude tables matching this regex'
  );
  outputOptions.option(
    '--include <regex>',
    'Include only tables matching this regex'
  );
  outputOptions.option('--silent', 'Suppress all console output');
  outputOptions.option('--debug', 'Enable debug logging');
  outputOptions.option(
    '--json-schema-import-location <path>',
    'Path to import JSON schemas'
  );
  outputOptions.addOption(
    new Option(
      '--zod-version <value>',
      'Zod version to use. (defaults to 3)'
    ).choices(['3', '4', '4-mini'])
  );
  outputOptions.addOption(
    new Option(
      '--case-transform <true|false>',
      'Whether to do case transformations / conversions for generated schemas. (defaults to true)'
    ).argParser(parseBooleanOption)
  );
  outputOptions.addOption(
    new Option(
      '--singularization <true|false>',
      'Whether to use singularization of type and enum names. (defaults to true)'
    ).argParser(parseBooleanOption)
  );
  outputOptions.addOption(
    new Option(
      '--coerce-dates <true|false>',
      'Whether to use z.coerce.date() for date fields in read schemas. (defaults to true)'
    ).argParser(parseBooleanOption)
  );
  outputOptions.addOption(
    new Option(
      '--stringify-json <true|false>',
      'Whether to JSON.stringify() on json fields in write schemas. (defaults to true)'
    ).argParser(parseBooleanOption)
  );
  outputOptions.addOption(
    new Option(
      '--stringify-dates <true|false>',
      'Whether to convert dates to ISO strings in write schemas. (defaults to false)'
    ).argParser(parseBooleanOption)
  );
  outputOptions.addOption(
    new Option(
      '--default-empty-array <true|false>',
      'Whether to use empty arrays as defaults for nullable array fields. (defaults to false)'
    ).argParser(parseBooleanOption)
  );
  outputOptions.addOption(
    new Option(
      '--default-unknown <true|false>',
      'Whether to use "unknown" instead of "any" for unresolved types. (defaults to false)'
    ).argParser(parseBooleanOption)
  );
  outputOptions.addOption(
    new Option(
      '--default-nulls-to-undefined <true|false>',
      'Whether to transform null values to undefined in generated read schemas. (defaults to true)'
    ).argParser(parseBooleanOption)
  );
  outputOptions.addOption(
    new Option(
      '--object-name-casing <type>',
      'Casing for generated object/type names. (defaults to PascalCase)'
    ).choices(['PascalCase', 'camelCase', 'snake_case'])
  );
  outputOptions.addOption(
    new Option(
      '--field-name-casing <type>',
      'Casing for field/property names in schemas & records. (defaults to camelCase)'
    ).choices(['camelCase', 'snake_case', 'PascalCase', 'passthrough'])
  );

  program.parse();
  const options = program.opts();
  const spinner = createProgressHandler(options.silent);

  try {
    const cliConfig: ZodDbsCliConfig = {
      ...config,
      ...options,
    };

    logDebug('Final configuration:', cliConfig);

    if (!cliConfig.silent) {
      logSettings(cliConfig);
      console.log();
    }

    await generateZodSchemas({
      provider,
      renderer: cliConfig.renderer,
      config: {
        ...cliConfig,
        onProgress: spinner.onProgress,
      },
    });

    spinner.done();
  } catch (error) {
    spinner.fail();

    logError(toError(error).message);
    logDebug(error);

    process.exit(1);
  }
};

const loadProvider = async (
  override: ZodDbsProvider | string | undefined,
  config: ZodDbsCliConfig
) => {
  try {
    const providerOrName =
      override ??
      getArgumentValue('-p') ??
      getArgumentValue('--provider') ??
      config.provider;

    return await importProvider(providerOrName);
  } catch (error) {
    logError(toError(error).message);
    logDebug(error);
    process.exit(1);
  }
};

const addProviderOptions = (program: Command, provider: ZodDbsProvider) => {
  if (!provider.options) return;

  const group = program.optionsGroup(
    `${provider.displayName ?? provider.name} options:`
  );

  for (const option of provider.options) {
    const flag =
      option.type === 'boolean'
        ? `--${option.name}`
        : `--${option.name} <value>`;
    const description = option.description;

    let cliOption = new Option(flag, description);

    if (option.allowedValues) {
      cliOption = cliOption.choices(option.allowedValues);
    }

    if (option.required) {
      cliOption = cliOption.makeOptionMandatory();
    }

    if (option.type === 'number') {
      cliOption = cliOption.argParser((value) => parseInt(value, 10));
    }

    group.addOption(cliOption);
  }
};

const logSettings = (config: ZodDbsCliConfig) => {
  for (const [key, value] of Object.entries(config)) {
    if (value !== undefined && value !== null) {
      logSetting(key, value);
    }
  }
};
