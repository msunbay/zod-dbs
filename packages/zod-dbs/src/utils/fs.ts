import { existsSync, readdirSync, unlinkSync } from 'node:fs';
import { mkdir } from 'node:fs/promises';
import path from 'node:path';
import { logDebug } from 'zod-dbs-core';

import type { ZodDbsTableType } from 'zod-dbs-core';

/**
 * Deletes all .ts files in the output schemas folder.
 */
export function clearTablesDirectory(outputPath: string) {
  // Check if the directory exists first
  if (!existsSync(outputPath)) {
    logDebug(`Directory ${outputPath} does not exist, nothing to clear`);
    return;
  }

  const files = readdirSync(outputPath, {
    recursive: true,
    withFileTypes: true,
  });

  for (const file of files) {
    if (file.isDirectory()) continue;
    if (!file.name.endsWith('.ts')) continue;

    const filePath = path.join(file.path, file.name);
    unlinkSync(filePath);
  }

  if (files.length > 0) {
    logDebug(`Deleted all .ts files in ${outputPath}`);
  }
}

export async function ensureFolder(folderPath: string) {
  const exists = existsSync(folderPath);

  if (!exists) {
    await mkdir(folderPath, { recursive: true });
  }
}

export const getOutputFolder = (type: ZodDbsTableType): string => {
  switch (type) {
    case 'table':
      return 'tables';
    case 'materialized_view':
      return 'materialized_views';
    case 'view':
      return 'views';
    case 'foreign_table':
      return 'foreign_tables';
    default:
      return 'others';
  }
};
