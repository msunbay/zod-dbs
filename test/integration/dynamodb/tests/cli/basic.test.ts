import fs from 'node:fs/promises';
import path from 'node:path';

import { executeCli, getOutputFiles } from '../../../utils/cli.js';
import { getProviderConfig } from '../../../utils/context.js';
import { getOutputDir } from '../../setup.js';

it('CLI generates correct zod schemas with basic options', async () => {
  const outputDir = getOutputDir('basic');
  const config = getProviderConfig();

  await executeCli(
    `--provider dynamodb --endpoint "${config.endpoint}" --access-key-id test --secret-access-key test --output-dir "${outputDir}" --silent --module-resolution esm`
  );

  const outputFiles = await getOutputFiles(outputDir);

  for (const file of outputFiles) {
    const content = await fs.readFile(file, 'utf8');
    expect(content).toMatchSnapshot(path.relative(outputDir, file));
  }
});
