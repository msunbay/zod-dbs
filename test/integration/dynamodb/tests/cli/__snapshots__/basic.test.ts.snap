// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CLI generates correct zod schemas with basic options > constants.ts 1`] = `
"// This file is auto-generated by zod-dbs. Do not edit this file directly.

export const TABLE_ORDERS = 'Orders';
export const TABLE_PRODUCTS = 'Products';
export const TABLE_SESSIONS = 'Sessions';
export const TABLE_USERS = 'Users';
"
`;

exports[`CLI generates correct zod schemas with basic options > tables/index.ts 1`] = `
"// This file is auto-generated by zod-dbs. Do not edit this file directly.

export * from './orders/index.js';
export * from './products/index.js';
export * from './sessions/index.js';
export * from './users/index.js';
"
`;

exports[`CLI generates correct zod schemas with basic options > tables/orders/index.ts 1`] = `
"// This file is auto-generated by zod-dbs. Do not edit this file directly.

export * from './schema.js';
"
`;

exports[`CLI generates correct zod schemas with basic options > tables/orders/schema.ts 1`] = `
"// This file is auto-generated by zod-dbs. Do not edit this file directly.

import { z } from 'zod';


/**
 * Read schema for the "Orders" table.
 * Validates raw rows read from the database.
 */
export const OrdersTableSchema = z.object({
     /**
      * dataType: string
      */
    createdAt: z.string(),
     /**
      * dataType: number
      */
    total: z.number(),
     /**
      * dataType: any
      */
    coupon: z.any(),
     /**
      * dataType: object
      */
    shipping: z.any(),
     /**
      * dataType: string
      */
    orderId: z.string(),
     /**
      * dataType: array
      */
    items: z.any(),
});

/**
 * Insert schema for the "Orders" table.
 */
export const OrdersTableInsertSchema = z.object({
     /**
      * dataType: string
      */
    createdAt: z.string(),
     /**
      * dataType: number
      */
    total: z.number(),
     /**
      * dataType: any
      */
    coupon: z.any(),
     /**
      * dataType: object
      */
    shipping: z.any(),
     /**
      * dataType: string
      */
    orderId: z.string(),
     /**
      * dataType: array
      */
    items: z.any(),
});

/**
 * Update schema for the "Orders" table.
 */
export const OrdersTableUpdateSchema = OrdersTableInsertSchema.partial();

type TableInsertRecord = z.input<typeof OrdersTableInsertSchema>;
type TableReadRecord = z.output<typeof OrdersTableSchema>;

/**
 * Read record for the ".Orders" table.
 */
export interface OrderRecord {
    /**
    */
    createdAt: TableReadRecord['createdAt'];
    /**
    */
    total: TableReadRecord['total'];
    /**
    */
    coupon: TableReadRecord['coupon'];
    /**
    */
    shipping: TableReadRecord['shipping'];
    /**
    */
    orderId: TableReadRecord['orderId'];
    /**
    */
    items: TableReadRecord['items'];
}

/**
 * Insertable record for the "Orders" table.
 */
export interface OrderInsertRecord {
    /**
    */
    createdAt: TableInsertRecord['createdAt'];
    /**
    */
    total: TableInsertRecord['total'];
    /**
    */
    coupon: TableInsertRecord['coupon'];
    /**
    */
    shipping: TableInsertRecord['shipping'];
    /**
    */
    orderId: TableInsertRecord['orderId'];
    /**
    */
    items: TableInsertRecord['items'];
}

/**
 * Updatable record for the "Orders" table.
 */
export type OrderUpdateRecord = Partial<OrderInsertRecord>;

"
`;

exports[`CLI generates correct zod schemas with basic options > tables/products/index.ts 1`] = `
"// This file is auto-generated by zod-dbs. Do not edit this file directly.

export * from './schema.js';
"
`;

exports[`CLI generates correct zod schemas with basic options > tables/products/schema.ts 1`] = `
"// This file is auto-generated by zod-dbs. Do not edit this file directly.

import { z } from 'zod';


/**
 * Read schema for the "Products" table.
 * Validates raw rows read from the database.
 */
export const ProductsTableSchema = z.object({
     /**
      * dataType: string
      */
    image: z.string(),
     /**
      * dataType: object
      */
    metadata: z.any(),
     /**
      * dataType: number
      */
    price: z.number(),
     /**
      * dataType: array
      */
    ratings: z.any(),
     /**
      * dataType: string
      */
    name: z.string(),
     /**
      * dataType: boolean
      */
    inStock: z.boolean(),
     /**
      * dataType: string
      */
    id: z.string(),
     /**
      * dataType: array
      */
    categories: z.any(),
});

/**
 * Insert schema for the "Products" table.
 */
export const ProductsTableInsertSchema = z.object({
     /**
      * dataType: string
      */
    image: z.string(),
     /**
      * dataType: object
      */
    metadata: z.any(),
     /**
      * dataType: number
      */
    price: z.number(),
     /**
      * dataType: array
      */
    ratings: z.any(),
     /**
      * dataType: string
      */
    name: z.string(),
     /**
      * dataType: boolean
      */
    inStock: z.boolean(),
     /**
      * dataType: string
      */
    id: z.string(),
     /**
      * dataType: array
      */
    categories: z.any(),
});

/**
 * Update schema for the "Products" table.
 */
export const ProductsTableUpdateSchema = ProductsTableInsertSchema.partial();

type TableInsertRecord = z.input<typeof ProductsTableInsertSchema>;
type TableReadRecord = z.output<typeof ProductsTableSchema>;

/**
 * Read record for the ".Products" table.
 */
export interface ProductRecord {
    /**
    */
    image: TableReadRecord['image'];
    /**
    */
    metadata: TableReadRecord['metadata'];
    /**
    */
    price: TableReadRecord['price'];
    /**
    */
    ratings: TableReadRecord['ratings'];
    /**
    */
    name: TableReadRecord['name'];
    /**
    */
    inStock: TableReadRecord['inStock'];
    /**
    */
    id: TableReadRecord['id'];
    /**
    */
    categories: TableReadRecord['categories'];
}

/**
 * Insertable record for the "Products" table.
 */
export interface ProductInsertRecord {
    /**
    */
    image: TableInsertRecord['image'];
    /**
    */
    metadata: TableInsertRecord['metadata'];
    /**
    */
    price: TableInsertRecord['price'];
    /**
    */
    ratings: TableInsertRecord['ratings'];
    /**
    */
    name: TableInsertRecord['name'];
    /**
    */
    inStock: TableInsertRecord['inStock'];
    /**
    */
    id: TableInsertRecord['id'];
    /**
    */
    categories: TableInsertRecord['categories'];
}

/**
 * Updatable record for the "Products" table.
 */
export type ProductUpdateRecord = Partial<ProductInsertRecord>;

"
`;

exports[`CLI generates correct zod schemas with basic options > tables/sessions/index.ts 1`] = `
"// This file is auto-generated by zod-dbs. Do not edit this file directly.

export * from './schema.js';
"
`;

exports[`CLI generates correct zod schemas with basic options > tables/sessions/schema.ts 1`] = `
"// This file is auto-generated by zod-dbs. Do not edit this file directly.

import { z } from 'zod';


/**
 * Read schema for the "Sessions" table.
 * Validates raw rows read from the database.
 */
export const SessionsTableSchema = z.object({
     /**
      * dataType: array
      */
    tokens: z.any(),
     /**
      * dataType: number
      */
    expires: z.number(),
     /**
      * dataType: string
      */
    sessionId: z.string(),
     /**
      * dataType: string
      */
    userId: z.string(),
});

/**
 * Insert schema for the "Sessions" table.
 */
export const SessionsTableInsertSchema = z.object({
     /**
      * dataType: array
      */
    tokens: z.any(),
     /**
      * dataType: number
      */
    expires: z.number(),
     /**
      * dataType: string
      */
    sessionId: z.string(),
     /**
      * dataType: string
      */
    userId: z.string(),
});

/**
 * Update schema for the "Sessions" table.
 */
export const SessionsTableUpdateSchema = SessionsTableInsertSchema.partial();

type TableInsertRecord = z.input<typeof SessionsTableInsertSchema>;
type TableReadRecord = z.output<typeof SessionsTableSchema>;

/**
 * Read record for the ".Sessions" table.
 */
export interface SessionRecord {
    /**
    */
    tokens: TableReadRecord['tokens'];
    /**
    */
    expires: TableReadRecord['expires'];
    /**
    */
    sessionId: TableReadRecord['sessionId'];
    /**
    */
    userId: TableReadRecord['userId'];
}

/**
 * Insertable record for the "Sessions" table.
 */
export interface SessionInsertRecord {
    /**
    */
    tokens: TableInsertRecord['tokens'];
    /**
    */
    expires: TableInsertRecord['expires'];
    /**
    */
    sessionId: TableInsertRecord['sessionId'];
    /**
    */
    userId: TableInsertRecord['userId'];
}

/**
 * Updatable record for the "Sessions" table.
 */
export type SessionUpdateRecord = Partial<SessionInsertRecord>;

"
`;

exports[`CLI generates correct zod schemas with basic options > tables/users/index.ts 1`] = `
"// This file is auto-generated by zod-dbs. Do not edit this file directly.

export * from './schema.js';
"
`;

exports[`CLI generates correct zod schemas with basic options > tables/users/schema.ts 1`] = `
"// This file is auto-generated by zod-dbs. Do not edit this file directly.

import { z } from 'zod';


/**
 * Read schema for the "Users" table.
 * Validates raw rows read from the database.
 */
export const UsersTableSchema = z.object({
     /**
      * dataType: string
      */
    id: z.string(),
     /**
      * dataType: string
      */
    sort: z.string(),
     /**
      * dataType: string
      */
    email: z.string(),
     /**
      * dataType: object
      */
    profile: z.any(),
     /**
      * dataType: array
      */
    tags: z.any(),
});

/**
 * Insert schema for the "Users" table.
 */
export const UsersTableInsertSchema = z.object({
     /**
      * dataType: string
      */
    id: z.string(),
     /**
      * dataType: string
      */
    sort: z.string(),
     /**
      * dataType: string
      */
    email: z.string(),
     /**
      * dataType: object
      */
    profile: z.any(),
     /**
      * dataType: array
      */
    tags: z.any(),
});

/**
 * Update schema for the "Users" table.
 */
export const UsersTableUpdateSchema = UsersTableInsertSchema.partial();

type TableInsertRecord = z.input<typeof UsersTableInsertSchema>;
type TableReadRecord = z.output<typeof UsersTableSchema>;

/**
 * Read record for the ".Users" table.
 */
export interface UserRecord {
    /**
    */
    id: TableReadRecord['id'];
    /**
    */
    sort: TableReadRecord['sort'];
    /**
    */
    email: TableReadRecord['email'];
    /**
    */
    profile: TableReadRecord['profile'];
    /**
    */
    tags: TableReadRecord['tags'];
}

/**
 * Insertable record for the "Users" table.
 */
export interface UserInsertRecord {
    /**
    */
    id: TableInsertRecord['id'];
    /**
    */
    sort: TableInsertRecord['sort'];
    /**
    */
    email: TableInsertRecord['email'];
    /**
    */
    profile: TableInsertRecord['profile'];
    /**
    */
    tags: TableInsertRecord['tags'];
}

/**
 * Updatable record for the "Users" table.
 */
export type UserUpdateRecord = Partial<UserInsertRecord>;

"
`;

exports[`CLI generates correct zod schemas with basic options > types.ts 1`] = `
"// This file is auto-generated by zod-dbs. Do not edit this file directly.

export type Table = 
    | 'Orders'
    | 'Products'
    | 'Sessions'
    | 'Users'
"
`;
